Add `stim.Circuit.{shortest,likeliest}_error_sat_problem`

noajshu opened the pull request:
## Summary
Adds a generator that outputs a .wcnf file string in [WDIMACS format](http://www.maxhs.org/docs/wdimacs.html).
Added a few tests as well.

## Demo for the surface code
Here is a demo:

**Step 1:** Generate the WCNF file
```python
import stim
c = stim.Circuit.generated(
            "surface_code:rotated_memory_z",
            distance=5,
            rounds=2,
            after_clifford_depolarization=0.001,
)
wcnf = c.shortest_error_sat_problem()
with open('d5problem.wcnf', 'w') as outfile:
    outfile.write(wcnf)
```

**Step 2:** Fetch a solver from the [2023 maxSAT competition](https://maxsat-evaluations.github.io/2023/)
```bash
wget https://maxsat-evaluations.github.io/2023/mse23-solver-src/exact/CASHWMaxSAT-CorePlus.zip
unzip CASHWMaxSAT-CorePlus.zip
```

**Step 3:** Run the solver on the produced file
```bash
time ./CASHWMaxSAT-CorePlus/bin/cashwmaxsatcoreplus -bm -m d5problem.wcnf 
c Parsing MaxSAT file...
c Using SCIP solver, version 8.0.3, https://www.scipopt.org
c scip_time = 0.000000
c Starting SCIP solver in a separate thread (with time-limit = 0s) ...
c Using COMiniSatPS SAT solver by Chanseok Oh (2016)
o 5
c ============================[  Problem Statistics ]============================
c |  Number of variables:           995                                         |
c |  Number of clauses:            4377 (incl.            5 soft in queue)      |
c ===============================================================================
c Optimal solution: 5
c _______________________________________________________________________________
c 
c restarts               : 881
c conflicts              : 88351          (20350 /sec)
c decisions              : 257865         (0.00 % random) (59394 /sec)
c propagations           : 6079046        (1400176 /sec)
c conflict literals      : 4980307        (6.84 % deleted)
c =======================[ UWrMaxSat resources usage ]===========================
c Memory used            : 164.00 MB
c CPU time               : 4.34163 s
c Wall clock time        : 3 s
c OptExp Enc: Srt/BDD/Add: 12 0 0
c _______________________________________________________________________________
v 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000110010100001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010100000100010000010010000000000001
s OPTIMUM FOUND

real	0m2.215s
user	0m4.345s
sys	0m0.051s
```

## Verifying distances of several quasicyclic codes
I ran this on some circuits from [this paper](https://arxiv.org/pdf/2308.07915.pdf), with help from @oscarhiggott to generate the circuits.
The log file is attached below:
[n72_n90_n103_quasicyclic_codes_have_expected_distances_6_8_8.txt](https://github.com/quantumlib/Stim/files/14427119/n72_n90_n103_quasicyclic_codes_have_expected_distances_6_8_8.txt)
So far, it has proven that the n=72, n=29, and n=103 circuits have the expected distance reported in [the paper](https://arxiv.org/pdf/2308.07915.pdf).
I am still running solvers on the n=144 and n=288 cases. My computer got rebooted so I had to start over. I am cautiously optimistic that the solvers will finish within a few weeks.

oscarhiggott commented on src/stim/search/sat/wcnf.cc:
```diff
@@ -0,0 +1,229 @@
+        }
+        // Add a soft clause for this error to be inactive
+        Clause clause;
+        clause.add_var(~err_x);
```
Since p>0.5 will lead to a disallowed negative weight, it might be worth handling the p>0.5 case with `clause.add_var(err_x);` and then flipping the sign of the weight?

oscarhiggott replied:
On second thoughts it may be best to leave it as is, since by default (unweighted case) the method finds the _shortest_ logical error, not the most probable.

noajshu replied:
Hmm, good catch!
I agree with both your first and second thoughts and I am troubled by the potential for accidental mistakes in using this.

It's unclear to me what the best choice of API is that balances utility with safety.

Do you think it might be appropriate to throw an InvalidArgument or cerr a warning if there is any error mechanism with probability >= 0.5? This way it's at least a visible error. WDYT?

oscarhiggott replied:
Yeah agree that just throwing an InvalidArgument about requiring p<0.5 is probably safest (and tbf it actually already does this with the message `"Clauses must have positive weight or HARD_CLAUSE_WEIGHT."` from line 58)

noajshu replied:
discussed offline and added another argument `weighted=False` where if it's True, does the p > 0.5 accounting Oscar suggested above.

Strilanc replied:
The docstring should probably explicitly say how p>0.5 errors are inverted if `weight_bins > 1`.

noajshu replied:
Done

Strilanc commented on src/stim/circuit/circuit.pybind.cc:
```diff
@@ -1966,6 +1971,45 @@ void stim_pybind::pybind_circuit_methods(pybind11::module &, pybind11::class_<Ci
+                ...     rounds=5,
+                ...     distance=5,
+                ...     after_clifford_depolarization=0.001)
+                >>> print(circuit.shortest_undetectable_logical_error_wcnf(
```
Use a smaller example, so that you can include the exact output.

noajshu replied:
Done

Strilanc commented on src/stim/circuit/circuit.pybind.cc:
```diff
@@ -1966,6 +1971,45 @@ void stim_pybind::pybind_circuit_methods(pybind11::module &, pybind11::class_<Ci
+            Generates a maxSAT problem instance in WDIMACS format whose optimal value is
+            the distance of the protocol, i.e. the minimum weight of any set of errors
+            that forms an undetectable logical error.
+
```
Include some notes on how you can use a publicly available solver to solve the result, like what you included in the PR message but shorter.

noajshu replied:
Done

Strilanc commented on src/stim/search/sat/wcnf.test.cc:
```diff
@@ -0,0 +1,55 @@
+  // hard clause x_0
+  // This gives a total of 8 clauses
+  // The top value should be at least 1 + 1 + 1 = 3. In our implementation ends up being 9.
+  std::stringstream expected;
```
Use a multiline string literal

```
expected = R"DIMACS(
so much space for activities!
)DIMACS";
```

Strilanc replied:
I don't understand why tabs in the data would prevent the use of multiline string literals.

Note that, for multiline literals, you don't indent the contents of the literal.

noajshu replied:
I'm going to push back against this suggestion again because it prevents adding explanatory inline comments between each of the lines.

Strilanc commented on src/stim/search/sat/wcnf.test.cc:
```diff
@@ -0,0 +1,55 @@
+
+TEST(shortest_undetectable_logical_error_wcnf, no_error) {
+    // No error.
+    ASSERT_THROW(
```
Is this the right behavior for the empty case? Shouldn't it return a file that embeds an unsatisfiable problem instead of failing, in this case?

Failing fast is great. I like failing fast. But I also like my code to work on the empty string when its semantics are compatible with the method. I'd be quite cross if `''.startswith('')` raised an exception that I was probably doing the wrong thing, instead of returning true.

noajshu replied:
We can certainly return an unsatisfiable wcnf in these cases.

Strilanc reviewed (commented):
Looks like you need to run the `regen_docs.sh` script in the `dev/` directory. It expects you to build and install the stim dev wheel before doing so (see the developer docs; this is easy to do with bazel).

noajshu commented on src/stim/circuit/circuit.pybind.cc:
```diff
@@ -1966,6 +1971,45 @@ void stim_pybind::pybind_circuit_methods(pybind11::module &, pybind11::class_<Ci
             .data());
+
+    c.def(
+        "shortest_undetectable_logical_error_wcnf",
```
Done

noajshu commented:
> Looks like you need to run the `regen_docs.sh` script in the `dev/` directory. It expects you to build and install the stim dev wheel before doing so (see the developer docs; this is easy to do with bazel).

Thanks, ran this and addressed all the comments.

Strilanc commented:
Note: the bug in `doctest_proper` is intended-behavior in `doctest`, where you have to say `<BLANKLINE>` instead of having blank output lines. But I prefer the modified print statements.

Strilanc commented on doc/python_api_reference_vDev.md:
```diff
@@ -2437,6 +2438,77 @@ def search_for_undetectable_logical_errors(
+    weighted: bool = False,
+    weight_scale_factor: int = 0,
+) -> str:
+    """Generates a maxSAT problem instance in WDIMACS format whose optimal value is
```
Stim's docstrings follow the convention that the first line is a self-contained paragraph summarizing the method.

I've made an attempt at improving the docstring:

```
Makes a maxSAT problem of the circuit's distance, that other tools can solve.

The output is a string describing the maxSAT problem in WDIMACS format
(see https://maxhs.org/docs/wdimacs.html). The optimal solution to the
problem is the fault distance of the circuit (the minimum number of error
mechanisms that combine to flip any logical observable while producing no
detection events).

There are many tools that can solve maxSAT problems in WDIMACS format.
For example, you can download one of the entries in the 2023 maxSAT
competition (see https://maxsat-evaluations.github.io/2023) and run it on
your problem by running these BASH terminal commands:

    wget https://maxsat-evaluations.github.io/2023/mse23-solver-src/exact/CASHWMaxSAT-CorePlus.zip
    unzip CASHWMaxSAT-CorePlus.zip
    ./CASHWMaxSAT-CorePlus/bin/cashwmaxsatcoreplus -bm -m your_problem.wcnf

Args:
    weight_bins: Defaults to 1 (unweighted errors). Error probabilities are converted to
        log-odds and scaled/rounded to be positive integers at most this large. Setting
        this argument to a larger number results in the solution to the problem being the
        most likely logical error weighted by probability of its parts, instead of just the
        one with the fewest parts.

Returns:
    A string corresponding to the contents of a WCNF file in WDIMACS format.

Examples:
    ...
```

noajshu replied:
Thanks. I had actually omitted the url because I got this error:
```
E   ImportError: Docstring line has length 98 > 80:
E       wget https://maxsat-evaluations.github.io/2023/mse23-solver-src/exact/CASHWMaxSAT-CorePlus.zip
E   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Any idea how to get past it?

Strilanc replied:
Go to https://github.com/quantumlib/Stim/blob/1cdf45f884442d01806d6a8997b867a352e937c9/src/stim/cmd/command_help.cc#L74 and add `"wget "` as a prefix string that disables the check. Or change it to allow if `https://` appears anywhere instead of just at the start.

noajshu replied:
Thank you!!

noajshu replied:
That doesn't work. I still get:
```
E   ImportError: Docstring line has length 98 > 80:
E       wget https://maxsat-evaluations.github.io/2023/mse23-solver-src/exact/CASHWMaxSAT-CorePlus.zip
E   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

noajshu replied:
I think the error is happening later on when we try to import the module, not when we clean the docstring

noajshu replied:
Ah nvm, I got it 🤦

Strilanc commented on doc/python_api_reference_vDev.md:
````diff
@@ -2437,6 +2438,77 @@ def search_for_undetectable_logical_errors(
+```
+
 <a name="stim.Circuit.shortest_graphlike_error"></a>
 ```python
````
This method should also be on `DetectorErrorModel`

noajshu replied:
That sounds nice, but could we leave that for a future PR after we've solidified the API here?
Also, note that the heuristic approach `search_for_undetectable_logical_errors` is not a method of `DetectorErrorModel`. It seems like if we add the SAT method to `DetectorErrorModel` then why not also add the heuristic method.
WDYT?

Strilanc commented on src/stim/search/sat/wcnf.cc:
```diff
@@ -0,0 +1,261 @@
+  size_t num_detectors = model.count_detectors();
+  size_t num_errors = model.count_errors();
+  if (num_observables == 0 or num_detectors == 0 or num_errors == 0) {
+    return UNSAT_WCNF_STR;
```
This is wrong. The no-det yes-err yes-obs case is often solvable with a result of d=1.

Add unit tests including a circuit with obs but no dets, dets but no obs, no obs and no dets, no errors, etc. Hit the corner cases.

Ideally, the method should not really check for the no-obs case, and instead just do what it normally does resulting in an unsatisfiable problem.

noajshu replied:
> This is wrong. The no-det yes-err yes-obs case is often solvable with a result of d=1.

good catch, fixed!

> Add unit tests including a circuit with obs but no dets, dets but no obs, no obs and no dets, no errors, etc. Hit the corner cases.

Done

> Ideally, the method should not really check for the no-obs case, and instead just do what it normally does resulting in an unsatisfiable problem.

I like your suggestion to 'do what it normally does'. Unfortunately SAT formulas do not permit clauses that involve constants which makes translating this edge case a bit unnatural to handle later in the translation.

If you would like aspects of the implementation to be changed I suggest we defer that to a future PR once the API and tests are solidified here.

Strilanc commented on src/stim/search/sat/wcnf.cc:
```diff
@@ -0,0 +1,261 @@
+      // For unweighted problems, all soft clauses have weight 1.
+      return 1;
+    }
+    return std::round(weight / max_weight * (double)weight_scale_factor);
```
This should be rounding such that it's never 0, right? It looks like this might end up as 0.

Add a unit test with an error of probability 49% as well as an error of probability 0.00000000001%.

noajshu replied:
I added these unit tests for the shortest error problem but for the likeliest error problem it's really supposed to round down to 0 in this case.

Strilanc replied:
Won't that result in instances that appear to be unsatisfiable, despite having a solution that's merely really-unlikely?

Strilanc commented on src/stim/search/sat/wcnf.cc:
```diff
@@ -0,0 +1,261 @@
+      // Each clause ends with 0
+      ss << " 0\n";
+    }
+    return ss.str();
```
Nit: I'd refactor the method into a `write_dimacs` style, taking a `std::ostream &out` parameter, instead of forcing the allocation of the string. Not necessary.

noajshu replied:
Ok, thanks.

Strilanc commented on src/stim/search/sat/wcnf.cc:
```diff
@@ -0,0 +1,261 @@
+  assert(error_index == num_errors);
+
+  // Add a hard clause for each detector to be inactive
+  for (size_t d=0; d<num_detectors; ++d) {
```
Nit: Run

```
find src | grep "\.\(cc\|h\|inl\)$" | grep -Pv "crumble_data.cc|gate_data_3d_texture_data.cc" | xargs clang-format -i
```

to autoformat the code.

noajshu replied:
thanks! done

Strilanc reviewed (changes requested):
Getting closer!

noajshu commented:
Hi @Strilanc , Oscar had an idea that I like for the interface and I was wondering what you thought.

The idea is to split into two python functions:

```python
minimum_cardinality_undetectable_logical_error_problem_as_maxsat_string(format)
approximately_most_probable_undetectable_logical_error_problem_as_maxsat_string(quantization_factor, format)
```

The idea would be to to ignore probabilities entirely in the first method and do a "best effort" approximation in the second method. "Best effort" means negation complementation and rounding to zero where it's the closest available integer.

WDYT

Strilanc commented:
I think these are close enough that they might as well be the same method.

Also those names are too long.

noajshu commented:
> I think these are close enough that they might as well be the same method with a flag of some kind.
> 
> Also those names are too long. If I was naming it I'd say `shortest_logical_error_as_...` and `most_likely_logical_error_as_...`.

I am happy to make either 2 methods with somewhat shorter names, as you suggest, or to keep it 1 method with a flag. Which solution do you prefer?

Strilanc commented:
Go with the two name solution.

noajshu commented on doc/python_api_reference_vDev.md:
```diff
@@ -46,6 +46,7 @@ API references for stable versions are kept on the [stim github wiki](https://gi
     - [`stim.Circuit.num_ticks`](#stim.Circuit.num_ticks)
     - [`stim.Circuit.reference_sample`](#stim.Circuit.reference_sample)
     - [`stim.Circuit.search_for_undetectable_logical_errors`](#stim.Circuit.search_for_undetectable_logical_errors)
+    - [`stim.Circuit.shortest_error_problem_as_wcnf_file`](#stim.Circuit.shortest_error_problem_as_wcnf_file)
```
done

noajshu commented on doc/python_api_reference_vDev.md:
```diff
@@ -2437,6 +2438,77 @@ def search_for_undetectable_logical_errors(
+        weighted: Defaults to False (unweighted), so that the problem is to find
+        the minimum size set of errors that leads to an undetectable logical
+        error.
+        weight_scale_factor: Defaults to 0. If weighted==False, must set
```
discussed offline, we now have 2 methods so this is out of date

noajshu commented on doc/python_api_reference_vDev.md:
```diff
@@ -2437,6 +2438,77 @@ def search_for_undetectable_logical_errors(
+
+    Args:
+        weighted: Defaults to False (unweighted), so that the problem is to find
+        the minimum size set of errors that leads to an undetectable logical
```
done

noajshu commented:
@Strilanc despite running this:
```
bazel build stim_dev_wheel
pip uninstall -y stim
pip install bazel-bin/stim-0.0.dev0-py3-none-any.whl
pytest src/stim/circuit/ && ./dev/regen_docs.sh
```
I don't see any updates to the docs. Yet, the CI job `test_generated_docs_are_fresh` does not pass. Do you have any idea what could be going wrong?

Strilanc commented:
The failure message i nthe CI log is

![image](https://github.com/quantumlib/Stim/assets/79941/83e29b16-19cf-4779-a8c4-b50b0adc9d15)

So it looks like your doctest code is invalid. The fresh-check also verifies the .pyi file runs as if it was python. You just need to make it be exactly valid (need '...' on every line including in multiline strings)

noajshu commented:
Thanks for the advice!
The problem with the docstring was simply that I used:
```
>>> circuit = stim.Circuit("""
...   X_ERROR(0.1) 0
...   M 0
...   OBSERVABLE_INCLUDE(0) rec[-1]
...   X_ERROR(0.4) 0
...   M 0
...   DETECTOR rec[-1] rec[-2]
... """)
```
which resulted in a """ inside a multiline string that itself used """.

I replaced these with ''' and also added a check to `clean_doc_string` to raise an error if there are any appearances of """ to help future fledgling Stim initiates like myself.

BTW -- it seems like some of the code in a recent commit to main was not formatted with this command you shared:
```
find src | grep "\.\(cc\|h\|inl\)$" | grep -Pv "crumble_data.cc|gate_data_3d_texture_data.cc" | xargs clang-format -i
```
you might consider running it on your branch.

Strilanc commented:
Because CI doesn't enforce that the code is formatted, it often drifts a bit over time.

viathor commented:
Cross-reference: This PR came up as an example to answer [this question](https://quantumcomputing.stackexchange.com/q/37289/10480) on QCSE.
